File: simulation/core/ethical_primes.py
"""
Ethical Primes Module

This module implements the concept of "ethical primes" - critical misjudgments
that represent fundamental errors in moral judgment systems.

It also computes Π(x), B(x), and E(x) functions analogous to the prime counting
function and error terms in number theory.
"""

import numpy as np
from typing import List, Tuple, Literal, Optional
from scipy.optimize import curve_fit
from .action_space import Action


def select_ethical_primes(
    actions: List[Action],
    importance_quantile: float = 0.9,
    strategy: Literal['importance', 'complexity', 'hybrid'] = 'importance',
    complexity_range: Optional[Tuple[int, int]] = None
) -> List[Action]:
    """
    Select "ethical primes" from the set of misjudged actions.
    
    Ethical primes are defined as critical misjudgments - cases that:
    1. Are actually misjudged (mistake_flag = 1)
    2. Have high importance (affect many people, high stakes)
    3. Fall within a meaningful complexity range (not trivial, not impossibly complex)
    
    Parameters
    ----------
    actions : List[Action]
        List of actions that have been judged
    importance_quantile : float, default=0.9
        Keep only top (1 - quantile) by importance
        E.g., 0.9 keeps top 10% most important
    strategy : {'importance', 'complexity', 'hybrid'}, default='importance'
        Selection strategy:
        - 'importance': Select by importance weight only
        - 'complexity': Prefer mid-range complexity
        - 'hybrid': Balance importance and complexity
    complexity_range : Optional[Tuple[int, int]], default=None
        If specified, only include actions with c in [min_c, max_c]
        If None, automatically set to exclude bottom 10% and top 10%
        
    Returns
    -------
    List[Action]
        Subset of actions identified as "ethical primes"
        
    Examples
    --------
    >>> actions = generate_world(1000)
    >>> judge = BiasedJudge()
    >>> evaluate_judgement(actions, judge, tau=0.3)
    >>> primes = select_ethical_primes(actions, importance_quantile=0.9)
    >>> print(f"Found {len(primes)} ethical primes")
    
    Notes
    -----
    The analogy with mathematical primes:
    - Not all misjudgments are "primes" - only the structurally important ones
    - Primes represent fundamental errors that can't be reduced to simpler cases
    - Their distribution tells us about the "health" of the judgment system
    """
    # Filter to only misjudgments
    mistakes = [a for a in actions if a.mistake_flag == 1]
    
    if len(mistakes) == 0:
        return []
    
    # Apply complexity range filter
    if complexity_range is not None:
        min_c, max_c = complexity_range
        mistakes = [a for a in mistakes if min_c <= a.c <= max_c]
    else:
        # Auto range: exclude bottom 10% and top 10% complexity
        all_complexities = sorted([a.c for a in mistakes])
        if len(all_complexities) >= 10:
            min_c = all_complexities[len(all_complexities) // 10]
            max_c = all_complexities[9 * len(all_complexities) // 10]
            mistakes = [a for a in mistakes if min_c <= a.c <= max_c]
    
    if len(mistakes) == 0:
        return []
    
    # Select by strategy
    if strategy == 'importance':
        # Sort by importance and take top quantile
        mistakes.sort(key=lambda a: a.w, reverse=True)
        cutoff_idx = int(len(mistakes) * (1 - importance_quantile))
        primes = mistakes[:max(cutoff_idx, 1)]
        
    elif strategy == 'complexity':
        # Prefer mid-range complexity (most interesting cases)
        # Score by distance from median complexity
        complexities = [a.c for a in mistakes]
        median_c = np.median(complexities)
        
        # Score: high importance, moderate complexity deviation
        def score(a):
            c_score = 1.0 / (1.0 + abs(a.c - median_c) / median_c)
            w_score = a.w / max(act.w for act in mistakes)
            return 0.3 * c_score + 0.7 * w_score
        
        mistakes.sort(key=score, reverse=True)
        cutoff_idx = int(len(mistakes) * (1 - importance_quantile))
        primes = mistakes[:max(cutoff_idx, 1)]
        
    elif strategy == 'hybrid':
        # Balance importance and error magnitude
        max_w = max(a.w for a in mistakes)
        max_delta = max(abs(a.delta) for a in mistakes) if mistakes else 1.0
        
        def score(a):
            w_normalized = a.w / max_w
            delta_normalized = abs(a.delta) / max_delta
            return 0.7 * w_normalized + 0.3 * delta_normalized
        
        mistakes.sort(key=score, reverse=True)
        cutoff_idx = int(len(mistakes) * (1 - importance_quantile))
        primes = mistakes[:max(cutoff_idx, 1)]
    else:
        raise ValueError(f"Unknown strategy: {strategy}")
    
    return primes


def compute_Pi_and_error(
    primes: List[Action],
    X_max: int = 100,
    baseline: Literal['linear', 'prime_theorem', 'fitted'] = 'prime_theorem',
    baseline_params: Optional[dict] = None
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Compute Π(x), B(x), and E(x) = Π(x) - B(x).
    
    This is the core function for the Ethical Riemann Hypothesis analog.
    
    Parameters
    ----------
    primes : List[Action]
        List of ethical primes
    X_max : int, default=100
        Maximum complexity to consider
    baseline : {'linear', 'prime_theorem', 'fitted'}, default='prime_theorem'
        Type of baseline function B(x):
        - 'linear': B(x) = α * x
        - 'prime_theorem': B(x) = β * x / log(x), analogous to Li(x)
        - 'fitted': Fit a smooth curve to Π(x)
    baseline_params : Optional[dict], default=None
        Parameters for baseline function
        - For 'linear': {'alpha': float}
        - For 'prime_theorem': {'beta': float}
        - For 'fitted': automatic fitting
        
    Returns
    -------
    Pi_x : np.ndarray
        Π(x) values - count of ethical primes up to complexity x
    B_x : np.ndarray
        B(x) values - baseline expectation
    E_x : np.ndarray
        E(x) values - error term E(x) = Π(x) - B(x)
    x_values : np.ndarray
        Array of x values from 1 to X_max
        
    Examples
    --------
    >>> primes = select_ethical_primes(actions)
    >>> Pi_x, B_x, E_x, x_vals = compute_Pi_and_error(primes, X_max=100)
    >>> import matplotlib.pyplot as plt
    >>> plt.plot(x_vals, E_x)
    >>> plt.xlabel('Complexity x')
    >>> plt.ylabel('Error E(x)')
    >>> plt.show()
    
    Notes
    -----
    The Ethical Riemann Hypothesis (ERH) states that:
    |E(x)| ≤ C * x^(1/2 + ε) for some constants C, ε > 0
    
    This function computes E(x) so we can test whether this bound holds.
    """
    x_values = np.arange(1, X_max + 1)
    Pi_x = np.zeros(X_max, dtype=float)
    
    # Compute Π(x): count of primes up to complexity x
    prime_complexities = [p.c for p in primes if p.c <= X_max]
    
    for i, x in enumerate(x_values):
        Pi_x[i] = sum(1 for c in prime_complexities if c <= x)
    
    # Compute baseline B(x)
    if baseline == 'linear':
        alpha = baseline_params.get('alpha', 0.1) if baseline_params else 0.1
        B_x = alpha * x_values
        
    elif baseline == 'prime_theorem':
        # B(x) = β * x / log(x), analogous to Prime Number Theorem
        beta = baseline_params.get('beta', 1.0) if baseline_params else 1.0
        
        # Avoid log(1) = 0
        B_x = np.zeros_like(x_values, dtype=float)
        for i, x in enumerate(x_values):
            if x > 1:
                B_x[i] = beta * x / np.log(x)
            else:
                B_x[i] = 0
                
    elif baseline == 'fitted':
        # Fit a smooth polynomial to Π(x)
        if len(primes) < 5:
            # Not enough data, fall back to linear
            B_x = 0.1 * x_values
        else:
            # Fit a degree-3 polynomial
            degree = min(3, len(primes) // 2)
            coeffs = np.polyfit(x_values, Pi_x, degree)
            B_x = np.polyval(coeffs, x_values)
    else:
        raise ValueError(f"Unknown baseline type: {baseline}")
    
    # Compute error
    E_x = Pi_x - B_x
    
    return Pi_x, B_x, E_x, x_values


def analyze_error_growth(
    E_x: np.ndarray,
    x_values: np.ndarray,
    expected_exponent: float = 0.5
) -> dict:
    """
    Analyze whether |E(x)| grows like x^α and estimate α.
    
    This tests the Ethical Riemann Hypothesis by checking if α ≈ 0.5.
    
    Parameters
    ----------
    E_x : np.ndarray
        Error values
    x_values : np.ndarray
        Complexity values
    expected_exponent : float, default=0.5
        Expected growth exponent (0.5 for ERH)
        
    Returns
    -------
    dict
        Analysis results including:
        - 'estimated_exponent': fitted α value
        - 'erh_satisfied': whether |α - 0.5| < 0.1
        - 'r_squared': goodness of fit
        - 'max_absolute_error': max |E(x)|
        - 'growth_rate': how E(x) grows
        
    Examples
    --------
    >>> Pi_x, B_x, E_x, x_vals = compute_Pi_and_error(primes)
    >>> analysis = analyze_error_growth(E_x, x_vals)
    >>> print(f"Estimated exponent: {analysis['estimated_exponent']:.3f}")
    >>> print(f"ERH satisfied: {analysis['erh_satisfied']}")
    """
    # Filter out zeros and take absolute value
    abs_E = np.abs(E_x)
    valid_mask = (abs_E > 0) & (x_values > 1)
    
    if np.sum(valid_mask) < 5:
        return {
            'estimated_exponent': np.nan,
            'erh_satisfied': False,
            'r_squared': 0.0,
            'max_absolute_error': np.max(abs_E) if len(abs_E) > 0 else 0,
            'growth_rate': 'insufficient_data'
        }
    
    x_valid = x_values[valid_mask]
    E_valid = abs_E[valid_mask]
    
    # Fit |E(x)| = C * x^α using log-log regression
    # log|E(x)| = log(C) + α * log(x)
    log_x = np.log(x_valid)
    log_E = np.log(E_valid)
    
    # Linear regression in log space
    coeffs = np.polyfit(log_x, log_E, 1)
    alpha = coeffs[0]  # slope = exponent
    log_C = coeffs[1]  # intercept = log(C)
    
    # Compute R² for goodness of fit
    log_E_pred = np.polyval(coeffs, log_x)
    ss_res = np.sum((log_E - log_E_pred)**2)
    ss_tot = np.sum((log_E - np.mean(log_E))**2)
    r_squared = 1 - (ss_res / ss_tot) if ss_tot > 0 else 0
    
    # Check if ERH is satisfied
    erh_satisfied = abs(alpha - expected_exponent) < 0.15
    
    # Classify growth rate
    if alpha < 0.4:
        growth_rate = 'sublinear_slow'  # Better than ERH!
    elif 0.4 <= alpha < 0.6:
        growth_rate = 'square_root'  # Consistent with ERH
    elif 0.6 <= alpha < 0.9:
        growth_rate = 'sublinear_fast'
    elif 0.9 <= alpha < 1.1:
        growth_rate = 'linear'
    else:
        growth_rate = 'superlinear'  # Problematic!
    
    return {
        'estimated_exponent': alpha,
        'constant_C': np.exp(log_C),
        'erh_satisfied': erh_satisfied,
        'r_squared': r_squared,
        'max_absolute_error': np.max(abs_E),
        'mean_absolute_error': np.mean(abs_E),
        'growth_rate': growth_rate,
        'deviation_from_erh': abs(alpha - expected_exponent)
    }


def compare_error_distributions(
    results: dict,
    X_max: int = 100
) -> dict:
    """
    Compare error distributions across multiple judges.
    
    Parameters
    ----------
    results : dict
        Dictionary mapping judge names to lists of evaluated actions
        (output from batch_evaluate)
    X_max : int, default=100
        Maximum complexity
        
    Returns
    -------
    dict
        Comparison results for each judge
    """
    comparison = {}
    
    for name, actions in results.items():
        primes = select_ethical_primes(actions)
        if len(primes) == 0:
            comparison[name] = {'error': 'no_primes'}
            continue
            
        Pi_x, B_x, E_x, x_vals = compute_Pi_and_error(primes, X_max=X_max)
        analysis = analyze_error_growth(E_x, x_vals)
        
        comparison[name] = {
            'num_primes': len(primes),
            'Pi_x': Pi_x,
            'B_x': B_x,
            'E_x': E_x,
            'x_values': x_vals,
            'analysis': analysis
        }
    
    return comparison


def compute_prime_density(
    primes: List[Action],
    X_max: int = 100,
    bin_size: int = 5
) -> Tuple[np.ndarray, np.ndarray]:
    """
    Compute the density of ethical primes in complexity bins.
    
    Parameters
    ----------
    primes : List[Action]
        List of ethical primes
    X_max : int, default=100
        Maximum complexity
    bin_size : int, default=5
        Size of complexity bins
        
    Returns
    -------
    bin_centers : np.ndarray
        Center of each bin
    densities : np.ndarray
        Number of primes per bin
    """
    bins = np.arange(0, X_max + bin_size, bin_size)
    bin_centers = (bins[:-1] + bins[1:]) / 2
    
    prime_complexities = [p.c for p in primes if p.c <= X_max]
    densities, _ = np.histogram(prime_complexities, bins=bins)
    
    return bin_centers, densities

File: julia/src/EthicalPrimes.jl
"""
    EthicalPrimes

Julia implementation of the ethical primes module.

Mirrors erh_core/core/ethical_primes.py.

"Ethical primes" are structurally-critical misjudgments — the subset of mistakes
that are most important, fall in a meaningful complexity range, and cannot be
decomposed into simpler cases.  The module also computes Π(x), B(x), and
E(x) = Π(x) − B(x), which are the core quantities for the Ethical Riemann
Hypothesis (ERH) analogy.
"""
module EthicalPrimes

using Statistics: median, mean, std, var
using LinearAlgebra: norm

export Action
export select_ethical_primes
export compute_Pi_and_error
export compute_error_correction_impact
export analyze_error_growth
export compute_prime_density

# ---------------------------------------------------------------------------
# Action struct — minimal analogue of the Python Action dataclass
# ---------------------------------------------------------------------------

"""
    Action

Minimal representation of a judged action.

Fields
------
- `c`            : complexity (integer ≥ 1)
- `w`            : importance weight (≥ 0)
- `V`            : ground-truth moral value
- `J`            : judge's assessment (nothing if not yet judged)
- `delta`        : error J − V (nothing if not judged)
- `mistake_flag` : 1 if |delta| > tau, 0 otherwise (nothing if not judged)
- `group`        : optional group label for fairness analysis
"""
mutable struct Action
    c            :: Int
    w            :: Float64
    V            :: Float64
    J            :: Union{Float64, Nothing}
    delta        :: Union{Float64, Nothing}
    mistake_flag :: Union{Int, Nothing}
    group        :: Union{String, Nothing}
end

# Convenience constructor — only the required fields.
Action(c::Int, w::Float64, V::Float64) =
    Action(c, w, V, nothing, nothing, nothing, nothing)

# ---------------------------------------------------------------------------
# select_ethical_primes
# ---------------------------------------------------------------------------

"""
    select_ethical_primes(actions; importance_quantile=0.9, strategy=:importance,
                          complexity_range=nothing) -> Vector{Action}

Select "ethical primes" from the set of misjudged actions.

Ethical primes are defined as critical misjudgments — cases that:
1. Are actually misjudged (`mistake_flag == 1`)
2. Have high importance (`w` above the given quantile)
3. Fall within a meaningful complexity range

Parameters
----------
- `importance_quantile` : keep only the top `(1 - quantile)` fraction by importance.
- `strategy`            : `:importance`, `:complexity`, or `:hybrid`.
- `complexity_range`    : `(min_c, max_c)` tuple; auto-computed if `nothing`.

Reference: Paper §3, Definition 3.1 (Ethical Primes).
"""
function select_ethical_primes(
    actions              :: Vector{Action};
    importance_quantile  :: Float64          = 0.9,
    strategy             :: Symbol           = :importance,
    complexity_range     :: Union{Tuple{Int,Int}, Nothing} = nothing,
) :: Vector{Action}

    # Filter to misjudgments
    mistakes = [a for a in actions if !isnothing(a.mistake_flag) && a.mistake_flag == 1]

    isempty(mistakes) && return Action[]

    # Apply complexity range filter
    if !isnothing(complexity_range)
        min_c, max_c = complexity_range
        mistakes = [a for a in mistakes if min_c <= a.c <= max_c]
    else
        # Auto range: exclude bottom 10 % and top 10 % complexity
        all_c = sort([a.c for a in mistakes])
        n = length(all_c)
        if n >= 10
            min_c = all_c[max(1, n ÷ 10 + 1)]
            max_c = all_c[min(n, 9 * n ÷ 10)]
            mistakes = [a for a in mistakes if min_c <= a.c <= max_c]
        end
    end

    isempty(mistakes) && return Action[]

    cutoff(lst) = max(1, round(Int, length(lst) * (1 - importance_quantile)))

    if strategy == :importance
        sort!(mistakes; by = a -> a.w, rev = true)
        return mistakes[1:cutoff(mistakes)]

    elseif strategy == :complexity
        complexities = Float64[a.c for a in mistakes]
        med_c = median(complexities)
        max_w = max(maximum(a.w for a in mistakes), 1e-9)

        score(a) = begin
            c_score = 1.0 / (1.0 + abs(a.c - med_c) / (med_c == 0 ? 1.0 : med_c))
            w_score = a.w / max_w
            0.3 * c_score + 0.7 * w_score
        end

        sort!(mistakes; by = score, rev = true)
        return mistakes[1:cutoff(mistakes)]

    elseif strategy == :hybrid
        max_w     = max(maximum(a.w for a in mistakes), 1e-9)
        max_delta = max(maximum(abs(a.delta === nothing ? 0.0 : a.delta) for a in mistakes), 1e-9)

        score(a) = begin
            w_n = a.w / max_w
            d_n = (isnothing(a.delta) ? 0.0 : abs(a.delta)) / max_delta
            0.7 * w_n + 0.3 * d_n
        end

        sort!(mistakes; by = score, rev = true)
        return mistakes[1:cutoff(mistakes)]

    else
        throw(ArgumentError("Unknown strategy: $strategy.  Use :importance, :complexity, or :hybrid."))
    end
end


# ---------------------------------------------------------------------------
# compute_Pi_and_error
# ---------------------------------------------------------------------------

"""
    compute_Pi_and_error(primes; X_max=100, baseline=:prime_theorem,
                         baseline_params=nothing)
          -> (Pi_x, B_x, E_x, x_values)

Compute Π(x), B(x), and E(x) = Π(x) − B(x).

Supported `baseline` symbols
-----------------------------
- `:linear`          — B(x) = α · x          (param: `alpha`, default 0.1)
- `:prime_theorem`   — B(x) = β · x / log(x) (param: `beta`,  default 1.0)
- `:power_law`       — B(x) = γ · x^δ        (params: `gamma`, `delta`)
- `:fitted`          — degree-3 polynomial fit to Π(x)

Reference: Paper §4, Definitions 4.1–4.3.
"""
function compute_Pi_and_error(
    primes         :: Vector{Action};
    X_max          :: Int            = 100,
    baseline       :: Symbol         = :prime_theorem,
    baseline_params :: Union{Dict, Nothing} = nothing,
) :: Tuple{Vector{Float64}, Vector{Float64}, Vector{Float64}, Vector{Int}}

    x_values = collect(1:X_max)

    # ------------------------------------------------------------------
    # Compute Π(x) : count of ethical primes with complexity ≤ x
    # ------------------------------------------------------------------
    prime_c = Float64[p.c for p in primes if p.c <= X_max]

    Pi_x = if isempty(prime_c)
        zeros(Float64, X_max)
    else
        # Vectorised: for each x, sum(c_i <= x)
        Float64[count(c -> c <= Float64(x), prime_c) for x in x_values]
    end

    # ------------------------------------------------------------------
    # Compute B(x)
    # ------------------------------------------------------------------
    params = isnothing(baseline_params) ? Dict{String,Float64}() : baseline_params

    B_x = if baseline == :linear
        alpha = get(params, "alpha", 0.1)
        alpha .* Float64.(x_values)

    elseif baseline == :prime_theorem
        beta = get(params, "beta", 1.0)
        # B(1) = 0  (log(1) = 0 → undefined → set to 0)
        [x == 1 ? 0.0 : beta * x / log(x) for x in x_values]

    elseif baseline == :power_law
        gamma = get(params, "gamma", 0.1)
        delta = get(params, "delta", 1.0)
        gamma .* Float64.(x_values) .^ delta

    elseif baseline == :fitted
        if length(primes) < 5
            0.1 .* Float64.(x_values)
        else
            degree = min(3, length(primes) ÷ 2)
            _polyfit_eval(Float64.(x_values), Pi_x, degree)
        end

    else
        throw(ArgumentError("Unknown baseline: $baseline.  Use :linear, :prime_theorem, :power_law, or :fitted."))
    end

    E_x = Pi_x .- B_x

    return Pi_x, B_x, E_x, x_values
end


# ---------------------------------------------------------------------------
# compute_error_correction_impact
# ---------------------------------------------------------------------------

"""
    compute_error_correction_impact(actions, mistake_indices; tau=0.3)
          -> Dict{Int, Float64}

Estimate the impact of correcting each mistake on the global error rate.

Reference: Paper §3 (structural fundamentality of a prime).
"""
function compute_error_correction_impact(
    actions         :: Vector{Action},
    mistake_indices :: Vector{Int};
    tau             :: Float64 = 0.3,
) :: Dict{Int, Float64}

    total_errors = count(a -> !isnothing(a.mistake_flag) && a.mistake_flag == 1, actions)
    n = length(actions)
    baseline_error_rate = n > 0 ? total_errors / n : 0.0

    impacts = Dict{Int, Float64}()

    for idx in mistake_indices
        idx > n && continue
        action = actions[idx]

        if isnothing(action.mistake_flag) || action.mistake_flag == 0
            impacts[idx] = 0.0
            continue
        end

        error_magnitude = isnothing(action.delta) ? 0.0 : abs(action.delta)

        # Count structurally similar mistakes
        similar_count = 0
        for other in actions
            other === action && continue
            (!isnothing(other.mistake_flag) && other.mistake_flag == 1) || continue

            c_denom  = max(action.c, other.c, 1)
            w_denom  = max(action.w, other.w, 0.001)
            c_diff   = abs(action.c - other.c) / c_denom
            w_diff   = abs(action.w - other.w) / w_denom

            if c_diff < 0.2 && w_diff < 0.2
                similar_count += 1
            end
        end

        estimated_reduction = action.w * error_magnitude * (1.0 + 0.1 * similar_count)

        impact = if baseline_error_rate > 0
            min(1.0, estimated_reduction / (baseline_error_rate * n))
        else
            0.0
        end

        impacts[idx] = impact
    end

    return impacts
end


# ---------------------------------------------------------------------------
# analyze_error_growth
# ---------------------------------------------------------------------------

"""
    analyze_error_growth(E_x, x_values; expected_exponent=0.5) -> Dict

Analyse whether |E(x)| grows like x^α and estimate α via log-log regression.

The ERH hypothesis predicts α ≈ 0.5.

Returns a `Dict` with keys:
- `estimated_exponent`  : fitted α
- `constant_C`          : fitted C in |E(x)| ≈ C · x^α
- `erh_satisfied`       : bool (bound-based, from `ERHChecks`)
- `r_squared`           : R² of the log-log fit
- `max_absolute_error`  : max |E(x)|
- `mean_absolute_error` : mean |E(x)|
- `growth_rate`         : qualitative label
- `deviation_from_erh`  : |α − 0.5|

Reference: Paper §6 (Error Growth Analysis).
"""
function analyze_error_growth(
    E_x              :: Vector{Float64},
    x_values         :: Vector{Int};
    expected_exponent :: Float64 = 0.5,
) :: Dict{String, Any}

    abs_E      = abs.(E_x)
    x_float    = Float64.(x_values)
    valid_mask = (abs_E .> 0) .& (x_float .> 1)

    max_abs_err  = isempty(abs_E) ? 0.0 : maximum(abs_E)
    mean_abs_err = isempty(abs_E) ? 0.0 : mean(abs_E)

    if count(valid_mask) < 5
        return Dict{String,Any}(
            "estimated_exponent"  => NaN,
            "constant_C"          => NaN,
            "erh_satisfied"       => false,
            "r_squared"           => 0.0,
            "max_absolute_error"  => max_abs_err,
            "mean_absolute_error" => mean_abs_err,
            "growth_rate"         => "insufficient_data",
            "deviation_from_erh"  => NaN,
            "erh_max_ratio"       => NaN,
            "erh_violation_rate"  => NaN,
        )
    end

    x_v = x_float[valid_mask]
    E_v = abs_E[valid_mask]

    log_x = log.(x_v)
    log_E = log.(E_v)

    # Linear regression in log space: log|E| = log(C) + α·log(x)
    alpha, log_C = _linear_regression(log_x, log_E)
    C = exp(log_C)

    # R² in log space
    log_E_pred = log_C .+ alpha .* log_x
    ss_res = sum((log_E .- log_E_pred) .^ 2)
    ss_tot = sum((log_E .- mean(log_E)) .^ 2)
    r2 = ss_tot > 0 ? 1.0 - ss_res / ss_tot : 0.0

    # ERH bound check (delegate to ERHChecks when available; inline here)
    bound_stats = _check_erh_bound_inline(E_x, x_float)

    deviation = abs(alpha - expected_exponent)

    growth_rate = if alpha < 0.4
        "sublinear_slow"
    elseif alpha < 0.6
        "square_root"
    elseif alpha < 0.9
        "sublinear_fast"
    elseif alpha < 1.1
        "linear"
    else
        "superlinear"
    end

    return Dict{String,Any}(
        "estimated_exponent"  => alpha,
        "constant_C"          => C,
        "erh_satisfied"       => bound_stats["erh_satisfied"],
        "r_squared"           => r2,
        "max_absolute_error"  => max_abs_err,
        "mean_absolute_error" => mean_abs_err,
        "growth_rate"         => growth_rate,
        "deviation_from_erh"  => deviation,
        "erh_max_ratio"       => bound_stats["max_ratio"],
        "erh_violation_rate"  => bound_stats["violation_rate"],
    )
end


# ---------------------------------------------------------------------------
# compute_prime_density
# ---------------------------------------------------------------------------

"""
    compute_prime_density(primes; X_max=100, bin_size=5)
          -> (bin_centers, densities)

Compute the density of ethical primes in complexity bins.
"""
function compute_prime_density(
    primes   :: Vector{Action};
    X_max    :: Int = 100,
    bin_size :: Int = 5,
) :: Tuple{Vector{Float64}, Vector{Int}}

    edges       = collect(0:bin_size:X_max)
    n_bins      = length(edges) - 1
    bin_centers = Float64[(edges[i] + edges[i+1]) / 2.0 for i in 1:n_bins]

    prime_c  = [p.c for p in primes if p.c <= X_max]
    densities = zeros(Int, n_bins)

    for c in prime_c
        for b in 1:n_bins
            if edges[b] < c <= edges[b+1]
                densities[b] += 1
                break
            end
        end
    end

    return bin_centers, densities
end


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

"""
Ordinary least squares: fit y = a·x + b, return (a, b).
"""
function _linear_regression(x::Vector{Float64}, y::Vector{Float64})
    n  = length(x)
    sx = sum(x);  sy = sum(y)
    sxx = sum(x .* x);  sxy = sum(x .* y)
    denom = n * sxx - sx^2
    denom == 0 && return (0.0, mean(y))
    a = (n * sxy - sx * sy) / denom
    b = (sy - a * sx) / n
    return a, b
end


"""
Inline ERH bound check (mirrors ERHChecks.check_erh_bound with default params).
Used inside EthicalPrimes to avoid a circular include dependency.
"""
function _check_erh_bound_inline(
    E_x      :: Vector{Float64},
    x_values :: Vector{Float64};
    C        :: Float64 = 1.0,
    epsilon  :: Float64 = 0.1,
    slack    :: Float64 = 1.5,
    avr      :: Float64 = 0.05,
) :: Dict{String, Any}

    valid = (x_values .> 0) .& isfinite.(E_x) .& isfinite.(x_values)
    if !any(valid)
        return Dict{String,Any}("erh_satisfied" => false,
                                "max_ratio"      => NaN,
                                "violation_rate" => NaN,
                                "num_points"     => 0)
    end

    x    = x_values[valid]
    absE = abs.(E_x[valid])
    bound  = C .* (x .^ (0.5 + epsilon))
    ratios = bound .> 0 ? absE ./ bound : fill(Inf, length(absE))

    max_ratio       = maximum(ratios)
    violation_rate  = mean(ratios .> 1.0)
    erh_satisfied   = (violation_rate <= avr) && (max_ratio <= slack)

    return Dict{String,Any}(
        "erh_satisfied"  => erh_satisfied,
        "max_ratio"      => max_ratio,
        "violation_rate" => violation_rate,
        "num_points"     => count(valid),
    )
end


"""
Fit a degree-`d` polynomial to `y` evaluated at `x`, return fitted values.
Uses normal equations via least squares (Vandermonde matrix).
"""
function _polyfit_eval(
    x :: Vector{Float64},
    y :: Vector{Float64},
    d :: Int,
) :: Vector{Float64}

    n = length(x)
    V = [x[i]^j for i in 1:n, j in 0:d]          # Vandermonde
    coeffs = V \ y                                  # least-squares solution
    return [sum(coeffs[j+1] * x[i]^j for j in 0:d) for i in 1:n]
end

end  # module EthicalPrimes
